### PR TITLE
ENH: Save to ZIP files without using temporary files.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
In Python 3.6 large data can be written to ZIP files directly, without using temporary files.

### CODE REVIEW

File: numpy/lib/npyio.py [hunk 1]
```diff
@@ -686,31 +686,41 @@ def _savez(file, args, kwds, compress, allow_pickle=True, pickle_kwargs=None):
...
-    os.close(fd)
-    try:
+    if sys.version_info >= (3, 6):
         for key, val in namedict.items():
             fname = key + '.npy'
-            fid = open(tmpfile, 'wb')
-            try:
-                format.write_array(fid, np.asanyarray(val),
+            val = np.asanyarray(val)
+            force_zip64 = val.nbytes > 1000000000
```
Comment (charris) [inline numpy/lib/npyio.py]: This might be clearer if a simple power of 2 is used, something like `val.nbytes > 2**30`.

File: numpy/lib/npyio.py [hunk 2]
```diff
@@ -686,31 +686,41 @@ def _savez(file, args, kwds, compress, allow_pickle=True, pickle_kwargs=None):
...
 
-    # Stage arrays in a temporary file on disk, before writing to zip.
-
-    # Since target file might be big enough to exceed capacity of a global
-    # temporary directory, create temp file side-by-side with the target file.
-    file_dir, file_prefix = os.path.split(file) if _is_string_like(file) else (None, 'tmp')
-    fd, tmpfile = tempfile.mkstemp(prefix=file_prefix, dir=file_dir, suffix='-numpy.npy')
-    os.close(fd)
-    try:
+    if sys.version_info >= (3, 6):
```
Comment (charris) [inline numpy/lib/npyio.py]: Might add a comment that since Python 3.6 it is possible to write directly to a zipfile.

**Comment by charris** [comment]
This is worthy of a release note. Could you add something to `doc/release/1.14.0-notes.rst`, probably under improvements.

**Comment by serhiy-storchaka** [comment]
Done. If you with I can add other improvements. Save small data without using temporary files even in Python <3.6. Use the context manager for avoiding file descriptor leak in case of failure.

**Comment by charris** [comment]
> Save small data without using temporary files even in Python < 3.6.

How would you do that? The problem is that the array data needs to be formatted and the resultant stream cannot be written directly to the zipfile.

**Comment by charris** [comment]
Upgrades are pretty easy these days, I'd be inclined to simply point to Python 3.6 for people who find the old implementation limiting.

**Comment by serhiy-storchaka** [comment]
The size of the formatted data can be estimated by the `nbytes` attribute of the array. If use in-memory files for formatting arrays smaller than say 1MB, this would significantly increase the performance of writing small arrays, but don't add too much to memory consumption (CPython itself needs tens megabytes to start).

**Comment by charris** [comment]
Hmm, those are pretty small files. I suspect speed is not of great importance here, but size can be. However, if you wish to add that I won't complain.

**Comment by serhiy-storchaka** [comment]
If speed is not of great importance here, I think it is not worth to complicate the code.

I have moved `import tempfile` which is not needed in 3.6.

File: numpy/lib/npyio.py [hunk 3]
```diff
@@ -686,31 +684,44 @@ def _savez(file, args, kwds, compress, allow_pickle=True, pickle_kwargs=None):
...
+        # Since Python 3.6 it is possible to write directly to a ZIP file.
         for key, val in namedict.items():
             fname = key + '.npy'
-            fid = open(tmpfile, 'wb')
-            try:
-                format.write_array(fid, np.asanyarray(val),
+            val = np.asanyarray(val)
+            force_zip64 = val.nbytes >= 2**30
+            with zipf.open(fname, 'w', force_zip64=force_zip64) as fid:
```
Comment (eric-wieser) [inline numpy/lib/npyio.py]: Is `force_zip64` an unrelated fix here?

Comment (eric-wieser) [reply numpy/lib/npyio.py]: and should this be `wb` to match the one below?

Comment (charris) [reply numpy/lib/npyio.py]: > The mode parameter, if included, must be 'r' (the default) or 'w'

I suspect the zip files are automatically binary.

Comment (serhiy-storchaka) [reply numpy/lib/npyio.py]: `force_zip64` is needed when write a stream. Tests are failed without this option.

`'r'` and `'w'` are the only supported options. Opened file-like objects are binary.

Comment (charris) [reply numpy/lib/npyio.py]: Looks like zip64 has been available since 2.7, but the `force_zip64` keyword is new. It is used for files of unknown size that may exceed 2 GiB. As we are formatting the array, I suppose it qualifies as of unknown size, but probably not by much.

**Comment by charris** [comment]
Thanks @serhiy-storchaka .

**Comment by serhiy-storchaka** [comment]
Thank you for your time and review @charris.

File: numpy/lib/npyio.py (hunk 3 above)
Comment (eric-wieser) [reply numpy/lib/npyio.py]: Should we be using `force_zip64` on all version of python 3, not just 3.6?

Comment (serhiy-storchaka) [reply numpy/lib/npyio.py]: `force_zip64` doesn't exist in earlier versions.
